x/cdp/keeper/interest.go

```diff
@@ -9,19 +9,80 @@ import (
 	"github.com/kava-labs/kava/x/cdp/types"
 )
 
-// CalculateFees returns the fees accumulated since fees were last calculated based on
-// the input amount of outstanding debt (principal) and the number of periods (seconds) that have passed
-func (k Keeper) CalculateFees(ctx sdk.Context, principal sdk.Coin, periods sdk.Int, collateralType string) sdk.Coin {
-	// how fees are calculated:
-	// feesAccumulated = (outstandingDebt * (feeRate^periods)) - outstandingDebt
-	// Note that since we can't do x^y using sdk.Decimal, we are converting to int and using RelativePow
-	feePerSecond := k.getFeeRate(ctx, collateralType)
-	scalar := sdk.NewInt(1000000000000000000)
-	feeRateInt := feePerSecond.Mul(sdk.NewDecFromInt(scalar)).TruncateInt()
-	accumulator := sdk.NewDecFromInt(types.RelativePow(feeRateInt, periods, scalar)).Mul(sdk.SmallestDec())
-	feesAccumulated := (sdk.NewDecFromInt(principal.Amount).Mul(accumulator)).Sub(sdk.NewDecFromInt(principal.Amount))
-	newFees := sdk.NewCoin(principal.Denom, feesAccumulated.TruncateInt())
-	return newFees
+var (
+	scalingFactor  = 1e18
+	secondsPerYear = 31536000
+)
+
+// AccumulateInterest calculates the new interest that has accrued for the input collateral type based on the total amount of principal
+// that has been created with that collateral type and the amount of time that has passed since interest was last accumulated
+func (k Keeper) AccumulateInterest(ctx sdk.Context, ctype string) error {
+	previousAccrualTime, found := k.GetPreviousAccrualTime(ctx, ctype)
+	if !found {
+		k.SetPreviousAccrualTime(ctx, ctype, ctx.BlockTime())
+		return nil
+	}
+
+	timeElapsed := ctx.BlockTime().Unix() - previousAccrualTime.Unix()
+	if timeElapsed == 0 {
+		return nil
+	}
+
+	totalPrincipalPrior := k.GetTotalPrincipal(ctx, ctype, types.DefaultStableDenom)
+	if totalPrincipalPrior.IsZero() || totalPrincipalPrior.IsNegative() {
+		return nil
+	}
+
+	interestFactorPrior, foundInterestFactorPrior := k.GetInterestFactor(ctx, ctype)
+	if !foundInterestFactorPrior {
+		interestFactorPrior = sdk.OneDec()
+		k.SetInterestFactor(ctx, ctype, interestFactorPrior)
+		// set previous accrual time exit early because interest accumulated will be zero
+		k.SetPreviousAccrualTime(ctx, ctype, ctx.BlockTime())
+		return nil
+	}
+
+	borrowRateSpy := k.getFeeRate(ctx, ctype)
+	if borrowRateSpy.Equal(sdk.OneDec()) {
+		return nil
+	}
+	interestFactor := CalculateInterestFactor(borrowRateSpy, sdk.NewInt(timeElapsed))
+	interestAccumulated := (interestFactor.Mul(totalPrincipalPrior.ToDec())).RoundInt().Sub(totalPrincipalPrior)
+	if interestAccumulated.IsZero() {
+		// in the case accumulated interest rounds to zero, exit early without updating accrual time
```

Comment:
Accural time might need to be updated every block?

Example: The fee rate is set to 1.00 by governance for some time. So `PreviousAccuralTime` isn't set for a while. Then governance sets the fee rate back to normal (>1). Then this func will run using the old accural time but using the new fee rate. It would calculate fees accruing over the period when the rate was 1.00 when no fees should accrue.

Reply:
Do param updates take effect before the BeginBlocker runs? If so, wouldn't any updates to the InterestFee param retroactively apply the new InterestFee rate to the last period? Such that when the fee rate is increased from 1.00 it'll still apply fees for a 1 block period for which the rate was 1.00 and when no fees should have accrued.

Reply:
Here I think we're trying to prevent the situation in which the 1-block accumulated interest rounds to 0 and no interest is applied over a long sequence of blocks during which the interest rate is > 1. Can we add a test that identifies the (probably low) interest rate/total principal amount thresholds at which this is a problem?

Reply:
Resolved in 80651c3. Handles the case where interest is rounded to zero and doesn't update `PreviousAccuralTime`. In cases where the APY is 0%, does update `PreviousAccuralTime`.